components/brave_mobile_subscription/renderer/android/subscription_render_frame_observer.cc

```diff
@@ -100,13 +104,87 @@ void SubscriptionRenderFrameObserver::DidCreateScriptContext(
 #endif
   } else if (product_ == Product::kLeo) {
     if (ai_chat_subscription_.is_bound()) {
+      AddJavaScriptObjectToFrame(context);
       ai_chat_subscription_->GetPurchaseTokenOrderId(base::BindOnce(
           &SubscriptionRenderFrameObserver::OnGetPurchaseTokenOrderId,
           weak_factory_.GetWeakPtr()));
     }
   }
 }
 
+void SubscriptionRenderFrameObserver::AddJavaScriptObjectToFrame(
+    v8::Local<v8::Context> context) {
+  if (!render_frame() || context.IsEmpty()) {
+    return;
+  }
+
+  v8::Isolate* isolate =
+      render_frame()->GetWebFrame()->GetAgentGroupScheduler()->Isolate();
+  v8::HandleScope handle_scope(isolate);
+  v8::Context::Scope context_scope(context);
+
+  CreateLinkResultObject(isolate, context);
+}
+
+void SubscriptionRenderFrameObserver::CreateLinkResultObject(
+    v8::Isolate* isolate,
+    v8::Local<v8::Context> context) {
+  v8::Local<v8::Object> global = context->Global();
+  v8::Local<v8::Object> link_result_obj;
+  v8::Local<v8::Value> link_result_value;
+  if (!global->Get(context, gin::StringToV8(isolate, "linkResult"))
+           .ToLocal(&link_result_value) ||
+      !link_result_value->IsObject()) {
+    link_result_obj = v8::Object::New(isolate);
+    global
+        ->Set(context, gin::StringToSymbol(isolate, "linkResult"),
+              link_result_obj)
+        .Check();
+    BindFunctionToObject(
+        isolate, link_result_obj, "setStatus",
+        base::BindRepeating(&SubscriptionRenderFrameObserver::SetLinkStatus,
+                            base::Unretained(this)));
+  }
+}
+
+template <typename Sig>
+void SubscriptionRenderFrameObserver::BindFunctionToObject(
+    v8::Isolate* isolate,
+    v8::Local<v8::Object> javascript_object,
+    const std::string& name,
+    const base::RepeatingCallback<Sig>& callback) {
+  v8::Local<v8::Context> context = isolate->GetCurrentContext();
+  // Get the isolate associated with this object.
+  javascript_object
+      ->Set(context, gin::StringToSymbol(isolate, name),
+            gin::CreateFunctionTemplate(isolate, callback)
+                ->GetFunction(context)
+                .ToLocalChecked())
+      .Check();
+}
+
+void SubscriptionRenderFrameObserver::SetLinkStatus(const std::string& status) {
```

Comment:
shouldn't this be `base::Value::Dict` to handle a js object (this might require adding a specialization to `brave/gin/converter_specializations.h`)? or js object is converted to `std::string` on call automatically?

or we expect the calling side do `JSON.stringify()`?